modules/ruxhal/src/platform/aarch64_common/ivc.rs

```diff
@@ -0,0 +1,218 @@
+     ivc_irqs: [u32; CONFIG_MAX_IVC_CONFIGS], // irq number
+ }
+ 
+ #[repr(C)]
+ #[derive(Debug)]
+ struct ControlTable {
+     ivc_id: u32,
+     max_peers: u32,
+     rw_sec_size: u32,
+     out_sec_size: u32,
+     peer_id: u32,
+     ipi_invoke: u32,
+ }
+ 
+ /// This module provides a way to establish a communication channel with a hypervisor (hvisor) 
+ /// for virtual machine (VM) communication using shared memory. Each VM can have two communication 
+ /// regions, and the region to be used for communication is specified during the connection setup.
+ ///
+ /// The communication is handled through the following steps:
+ /// 
+ /// 1. **Connection Setup**: The `connect()` function allocates memory for communication structures
+ ///    and invokes a hypervisor call (`hvc`) to retrieve necessary information about the IVC 
+ ///    (Inter-VM Communication) and control tables. The specific communication region to be used 
+ ///    is determined by the parameter passed to `connect()`. This process sets up the shared memory 
+ ///    and prepares the communication channel.
+ /// 2. **Message Sending**: The `send_message()` function writes a message to the shared memory area 
+ ///    specified by the hypervisor. The message is written to a predefined memory location, and 
+ ///    the control table is updated to notify the target VM of the message.
+ /// 3. **Connection Teardown**: The `close()` function frees the allocated memory and closes the 
+ ///    communication channel, cleaning up resources to prevent memory leaks.
+ ///
+ /// The module relies on a `GlobalAllocator` for memory management and uses raw pointers and unsafe 
+ /// Rust operations to interact with memory addresses provided by the hypervisor. It is critical 
+ /// that the communication sequence follows the correct order: connect -> send_message -> close.
+ ///
+ /// # Example
+ /// ```
+ /// let mut conn = Connection::new();
+ /// if let Err(e) = conn.connect(0) {    // Choose the first communication region (0)
+ ///     info!("Error connecting: {}", e);
+ ///     return;
+ /// }
+ /// if let Err(e) = conn.send_message("Hello from VM!") {
+ ///     error!("Error sending message: {}", e);
+ /// }
+ /// if let Err(e) = conn.close() {
+ ///     error!("Error closing connection: {}", e);
+ /// }
+ /// ```
+ 
+ pub fn ivc_example() {
+     let mut conn = Connection::new();
+ 
+     // Establish the connection
+     if let Err(e) = conn.connect(0) {
+         info!("Error connecting: {}", e);
+         return;
+     }
+ 
+     // Send the message
+     if let Err(e) = conn.send_message("Hello from VM!") {
+         info!("Error sending message: {}", e);
+     }
+ 
+     // Close the connection
+     if let Err(e) = conn.close() {
+         info!("Error closing connection: {}", e);
+     }
+ }
+ 
+ pub struct Connection<'a> {
+     // Reference to the allocator
+     allocator: &'a GlobalAllocator, 
+     ivc_info_ptr: *mut IvCInfo,
```

Comment:
`ivc_info_ptr` can be turned into a safer type, try `SpinLock<Option<IvCInfo>` or `IvCInfo` directly.

Reply:
fixed.